File: solutions/julia/complex-numbers/complex-numbers.jl
enable_realcomplex_tests = true
enable_syntaxsugar_tests = true

import Base: +, -, *, /, ==, ≈, conj, abs, exp, real, imag

struct ComplexNumber <: Number
    r
    i
    ComplexNumber(r, i) = new(r, i)
    ComplexNumber(r) = ComplexNumber(r, 0)
end

const jm = ComplexNumber(0, 1)

+(z1::ComplexNumber, z2::ComplexNumber) = ComplexNumber(z1.r + z2.r, z1.i + z2.i)
+(z1::ComplexNumber, z2::Real) = z1 + ComplexNumber(z2, 0)
+(z1::Real, z2::ComplexNumber) = ComplexNumber(z1, 0) + z2
-(z1::ComplexNumber, z2::ComplexNumber) = ComplexNumber(z1.r - z2.r, z1.i - z2.i)
-(z1::ComplexNumber, z2::Real) = z1 - ComplexNumber(z2, 0)
-(z1::Real, z2::ComplexNumber) = ComplexNumber(z1, 0) - z2
*(z1::ComplexNumber, z2::ComplexNumber) = ComplexNumber(z1.r * z2.r - z1.i * z2.i, z1.r * z2.i + z1.i * z2.r)
*(z1::ComplexNumber, z2::Real) = z1 * ComplexNumber(z2, 0)
*(z1::Real, z2::ComplexNumber) = ComplexNumber(z1, 0) * z2
/(z1::ComplexNumber, z2::ComplexNumber) = ComplexNumber((z1.r * z2.r + z1.i * z2.i) / abs(z2)^2, (-z1.r * z2.i + z1.i * z2.r) / abs(z2^2))
/(z1::ComplexNumber, z2::Real) = ComplexNumber(z1.r / z2, z1.i / z2)
/(z1::Real, z2::ComplexNumber) = ComplexNumber((z1 * z2.r) / (z2.r^2 + z2.i^2), (-z1 * z2.i) / (z2.r^2 + z2.i^2))
==(z1::ComplexNumber, z2::ComplexNumber) = z1.r == z2.r && z1.i == z2.i
≈(z1::ComplexNumber, z2::ComplexNumber) = z1.r ≈ z2.r && z1.i ≈ z2.i

abs(z::ComplexNumber) = sqrt(z.r^2 + z.i^2)
conj(z::ComplexNumber) = ComplexNumber(z.r, -z.i)
exp(z::ComplexNumber) = ComplexNumber(exp(z.r) * cos(z.i), exp(z.r) * sin(z.i))
real(z::ComplexNumber) = z.r
imag(z::ComplexNumber) = z.i

File: solutions/cpp/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"

#include <cmath>

namespace complex_numbers {


    Complex::Complex(double r) : m_r(r), m_i(0) {};
    Complex::Complex(double r, double i) : m_r(r), m_i(i) {}

    auto Complex::real() const -> double { return m_r; }
    auto Complex::imag() const -> double { return m_i; }

    auto Complex::conj() const -> Complex { return Complex(m_r, -m_i); }
    auto Complex::abs() const  -> double { return std::sqrt(m_r * m_r + m_i * m_i); };

    auto operator+(const Complex &z1, const Complex &z2) -> Complex {
        return Complex(z1.real() + z2.real(), z1.imag() + z2.imag());
    }

    auto operator-(const Complex &z1, const Complex &z2) -> Complex {
        return Complex(z1.real() - z2.real(), z1.imag() - z2.imag());
    }

    auto operator*(const Complex &z1, const Complex &z2) -> Complex {
        return Complex (
            z1.real() * z2.real() - z1.imag() * z2.imag(),
            z1.real() * z2.imag() + z1.imag() * z2.real()
        );
    }

    auto operator/(const Complex &z1, const Complex &z2) -> Complex {
        return Complex (
           (z1 * z2.conj()).real() / (z2.abs() * z2.abs()),
           (z1 * z2.conj()).imag() / (z2.abs() * z2.abs())
        );
    }

    auto Complex::exp() const -> Complex {
        return Complex (
            std::exp(m_r) * std::cos(m_i),
            std::exp(m_r) * std::sin(m_i)
        );
    }

}  // namespace complex_numbers
